apiserver/facades/client/applicationoffers/applicationoffers.go

```diff
@@ -240,75 +279,69 @@ func (api *OffersAPIv5) modifyOneOfferAccess(ctx context.Context, user names.Use
 	if err != nil {
 		return errors.Annotate(err, "could not modify offer access")
 	}
-	return api.changeOfferAccess(backend, url.ApplicationName, targetUserTag, arg.Action, offerAccess)
+	return api.changeOfferAccess(ctx, backend, url.ApplicationName, targetUserTag, arg.Action, offerAccess)
 }
 
 // changeOfferAccess performs the requested access grant or revoke action for the
 // specified user on the specified application offer.
 func (api *OffersAPIv5) changeOfferAccess(
+	ctx context.Context,
 	backend Backend,
 	offerName string,
 	targetUserTag names.UserTag,
 	action params.OfferAction,
-	access permission.Access,
+	accessLevel permission.Access,
 ) error {
+	targetUserName := coreuser.NameFromTag(targetUserTag)
+	apiUserName := coreuser.NameFromTag(api.Authorizer.GetAuthTag().(names.UserTag))
+
 	offer, err := backend.ApplicationOffer(offerName)
 	if err != nil {
 		return errors.Trace(err)
 	}
 	offerTag := names.NewApplicationOfferTag(offer.OfferUUID)
+
+	var change permission.AccessChange
 	switch action {
 	case params.GrantOfferAccess:
-		return api.grantOfferAccess(backend, offerTag, targetUserTag, access)
+		change = permission.Grant
 	case params.RevokeOfferAccess:
-		return api.revokeOfferAccess(backend, offerTag, targetUserTag, access)
+		change = permission.Revoke
 	default:
 		return errors.Errorf("unknown action %q", action)
 	}
-}
-
-func (api *OffersAPIv5) grantOfferAccess(backend Backend, offerTag names.ApplicationOfferTag, targetUserTag names.UserTag, access permission.Access) error {
-	err := backend.CreateOfferAccess(offerTag, targetUserTag, access)
-	if errors.IsAlreadyExists(err) {
-		offerAccess, err := backend.GetOfferAccess(offerTag.Id(), targetUserTag)
-		if errors.IsNotFound(err) {
-			// Conflicts with prior check, must be inconsistent state.
-			err = jujutxn.ErrExcessiveContention
-		}
-		if err != nil {
-			return errors.Annotate(err, "could not look up offer access for user")
-		}
 
-		// Only set access if greater access is being granted.
-		if offerAccess.EqualOrGreaterOfferAccessThan(access) {
-			return errors.Errorf("user already has %q access or greater", access)
-		}
-		if err = backend.UpdateOfferAccess(offerTag, targetUserTag, access); err != nil {
-			return errors.Annotate(err, "could not set offer access for user")
-		}
-		return nil
+	var addUser bool
+	var external *bool
+	if ext := !targetUserName.IsLocal(); ext {
+		addUser = true
+		external = &ext
```

Comment:
add a comment that addUser is only done for external to be clear. I'm a bit surprised that this is being added as the prior logic did not.

Reply:
Previously there were no external users stored in mongo at all, they only existed in permissions. Now with referential integrity we need to link to something in the users table. Its been a bit of a bodge on to be honest, we've ended up with methods like this where we add external users as we discover them. 

One of those places is `ReadUserAccessLevelForTargetAddingMissingUser` but they can be added via a grant before they log in too, i.e. to do `juju grant aflynn50@external read offerer.offer` we will need to create the user aflynn50, which is what we are doing here.